modules/auxiliary/scanner/scada/modbus_zip.rb

```diff
@@ -0,0 +1,78 @@
+##
+## This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+require 'packetfu'
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Exploit::Capture
+
+  def initialize(info = {})
+    super(
+      'Name'           => "Project File",
+      'Description'    => %q{
+        This module is able to extract a zip file sent through Modbus from a pcap.
+        Tested with Schneider TM221CE16R
+      },
+        'Author' => [
+          'José Diogo Monteiro <jdlopes[at]student.dei.uc.pt>',
+          'Luis Rosa <lmrosa[at]dei.uc.pt)>',
+        ],
+      'License' => MSF_LICENSE
+      )
+    register_options([Opt::RPORT(502), Opt::RHOSTS("0.0.0.0"),
+                      OptEnum.new("MODE", [true, 'Extract zip from upload/download capture', "UPLOAD",
+                                           [
+                                             "UPLOAD",
+                                             "DOWNLOAD"
+                                           ]]),
+                      OptString.new('FILENAME', [ true, "The zip file", File.join(Msf::Config.module_directory, "auxiliary", "scanner", "scada", "projectfile.zip") ]),])
+  end
+
+  def extract_zip(packet, zip_packet, first_byte, file, packet_number)
```

Comment:
Would it be possible to have the start and end signatures in the same packet?

Reply:
Maybe, if the file is small enough (empty project)
However, the main use case consists of extracting a projectfile from a PLC (which is large enough)

Reply:
Hmm, okay.